### ISSUE TITLE
debounced, deferred or throttled calls

### ISSUE THREAD

**Issue by konsultaner** [wamp-proto/wamp-proto#391]
I sometimes but rarely have an issue where a caller invokes a procedure twice within a very little time interval. This is usually no issue. I do have a scenario where I need to make sure that the first call is fully processed before the next call enters the business logic. The problem only occurs when the procedure is called with the same kwargs or args resp. I've been thinking about this a lot and it seems like adding a debounce to wamp seems a better solution then handling this issue within the business logic.

There are some problems I immediately thought of:

- who handles the debounce
- who is allowed to decide wether to debounce or not
- are there issues with memory overflow at some point
- how to response to the client

I kind of came up with this:

A callee registers a procedure with an option `processing_control=debounce`. The router will then allow any caller to add an `transaction_hash=???` to the call options. The hash is generated be the caller and lets the caller decide which fields must be unique to the call. If two callers send the same `transaction_hash` at the same time, the router returns an `abort` for one of that call depending on the `processing_control`.

A call could look like this: 

`[48, 7814135, {"transaction_hash":"F4EB27CEA7255CEA4D1FFABF593372E8"}, "com.myapp.user.new", ["johnny"], {"firstname": "John", "surname": "Doe"}]`

The hash was generated by `MD5("F4EB27CEA7255CEA4D1FFABF593372E8")`

This means any other call with this hash will be aborted or deferred (depending on how its implemented). Since the caller would always generate this hash for any call having `args[0] == "johnny"`, user interaction, network issues or whatever would always block douple execution.

Why would I not let the router decide when to debounce?

1. opaque payload may make it impossible
2. the router would need to store extra state

Why would I not let the callee decide when to debounce?

1. if there are multiple callees registred on the same procedure they would have to sync states
2. In my case there are multiple callees but one database -> without syncing the states multiple entries may appear

Looking forward to the feedback!

**Comment by oberstet** [comment]
>  I do have a scenario where I need to make sure that the first call is fully processed before the next call enters the business logic. 

you can do that (in Crossbar.io at least) using *concurrency control*

```
yield self.register(self.compute,
                    u'com.example.compute_serialized',
                    options=RegisterOptions(concurrency=1))
```

you can also specify *call queuing* to control how many unprocessed calls will be queued in the router

https://github.com/crossbario/crossbar-examples/tree/master/scaling-microservices

**Comment by oberstet** [comment]
> if there are multiple callees registred on the same procedure they would have to sync states

if you have `concurrency=1`, but still have multiple callees (all with that same concurrency specified), then at most 1 call will be processed concurrently *for the whole router node* - but the specific callee processing one given call will go through your list of callees

obviously, this makes limited sense

if you need to coordinate between different callees concurrently processing multiple calls in other ways, then you need to do that in your callees.

> you can create hard to debug scenarios easily using locking in a shared DB ... deadlocks. if you go down that route, make sure you have a monitor running checking the DB for deadlocks all the time, and then actively kill DB sessions or WAMP components (callee instances) when needed

**Comment by oberstet** [comment]
yet another note rgd Crossbar.io and concurrency control: currently, this is implemented at the caller originating router node, and hence, in multi-node scenarios (eg you have a full mesh of 10 Crossbar.io nodes), the actual concurrency arriving at the callee can exceed the specified concurrency - as calls can _originate_ from multiple nodes

> @om26er ^ as discussed, the r2r testing for _all_ the features of Crossbar.io is a challenge;) we have a lot of features. I did expect this .. WAMP was designed to be able to freely combine features .. and this is non trivial implementation wise. anyways, just a note because we had it just now ..

**Comment by konsultaner** [comment]
@oberstet I see what you mean, but I do not need it per procedure, but per procedure with certain kwargs or args. 

Lets say I have a procedure `order.a.meal` called with kwargs `{amount:1, date:25.01.2021, mealId:129302}`. I don't care if another call is made in parallel as long as the fields `date` and `mealId` are not the same.

This is ok:

```
[48, 7814135, {}, "order.a.meal", [], `{amount:1, date:25.01.2021, mealId:129302}]
[48, 7814135, {}, "order.a.meal", [], `{amount:1, date:26.01.2021, mealId:345302}]
```

But this should cancel out the second call:

```
[48, 7814135, {}, "order.a.meal", [], `{amount:1, date:25.01.2021, mealId:129302}]
[48, 7814135, {}, "order.a.meal", [], `{amount:1, date:25.01.2021, mealId:129302}] <-- don't execute
```

That is why I thought of adding a hash on the caller side to prevent all of this:

>Why would I not let the router decide when to debounce?
>
>1. opaque payload may make it impossible
>2. the router would need to store extra state
>
>Why would I not let the callee decide when to debounce?
>
>1. if there are multiple callees registred on the same procedure they would have to sync states
>2. In my case there are multiple callees but one database -> without syncing the states multiple entries may appear

I don't work on IoT projects with wamp. Thats why it may seem a bit of an off scenario. But for non IoT project this may be an issue for a lot of wamp users.

**Comment by oberstet** [comment]
ahh, ok, right: what you want is different from the features I posted.

it is really: automatically and silently drop calls to the same procedure with the same args & kwargs in the router.

so the duplicate call is issued by the caller, but recognized by the router. the router then can drop the callee invocation and return an error or the result that had been returned for the first call with respective args/kwargs.

if this is what you want, it requires keeping state of at least the hash args&kwargs per procedure URI - and optionally also for the result of such call

is ^ more what you thought of? 

**auto-deduplication of calls/publishes in router based on hash(args+kwargs ) on per URI basis**

**Comment by konsultaner** [comment]
This is exactly what I want! What do you think about it? And my current solution?

**Comment by oberstet** [comment]
re-reading - now in the light of that I actually understand what's the goal - your proposal seems better actually

> callee registers a procedure with an option processing_control=debounce. The router will then allow any caller to add an transaction_hash=??? to the call options. The hash is generated be the caller and lets the caller decide which fields must be unique to the call. If two callers send the same transaction_hash at the same time, the router returns an abort for one of that call depending on the processing_control.

this is a good starting point .. we'll need to figure out more .. couple of first notes

* "who is allowed to decide wether to debounce" => the URI must be configure to allow auto-dedup at router-side (under admin control) => if that is enabled, a client MAY then request auto-dedup when registering
* "are there issues with memory overflow at some point" => yes, it consumes mem, and must be bounded hence. => "who configures the memory depth of the auto-dedup"
* will it only work for RPCs, or also for PubSub? eg auto-dedup on publish
* do we want a new option processing_control=debounce, or could we use a new mode value in an existing option ..
* "debounce": an alternative word would be "deduplication" and "auto-dedup" or similar. deduplication is what data scientists know ... debouncing is sth eletronics guys do;)
* "transaction_hash": it needs to be unique .. "hash" suggests that, but also suggest a specific way of computing the value (in the client lib or app) .. which is unnecessary. and "ID" is good enough ..
* rgd return policy (that is, upon detecting a dup): "return error", "return first", ..?

**Comment by konsultaner** [comment]
> this is a good starting point .. we'll need to figure out more .. couple of first notes

I thought so. But I thought I'd just start with something.

> "who is allowed to decide wether to debounce" => the URI must be configure to allow auto-dedup at router-side (under admin control) => if that is enabled, a client MAY then request auto-dedup when registering

IMO dedup is somehow related to `invocation_policy`. So to me it seems that allowing to dedup is done on the callee side. The callee knows about the business logic and also knows wether it may be necessary or not.
What is security wise the worst case scenario: The router will cancel all incoming calls. This doesn't seem to be very insecure? The only thing that may happen is an out of memory, Which may be controlled by a max_hashes (or max_ids) configuration on the routers side. I think this is specific to the implementers router and should not be part of the spec.

> "are there issues with memory overflow at some point" => yes, it consumes mem, and must be bounded hence. => "who configures the memory depth of the auto-dedup"

As I said, I think this should be handled individually by the implementers. The only thing that needs to be stored is a list of hashes per procedure. Any hash will be removed after the call was processed. So invalidating a spammers session or setting a max_hash limit will work. The hash stack will eventually clean up by itself.

> will it only work for RPCs, or also for PubSub? eg auto-dedup on publish

Since pub/sub has no response it would make it a bit harder. The router would not know when to invalidate the hash. This would only work with some timeout. IMO timeouts are hard to estimate in a distributed system and thus cause some unwanted behaviour. Especially when they are not in a local network. What do I usually use pub/sub for? Only for synchronization. I hardly ever use it so compute something or activate business logic. I don't see a scenario for that.

> do we want a new option processing_control=debounce, or could we use a new mode value in an existing option

This is something independent and may be used in conjuctions with all the other features that we have. Thinking about it a little more maybe the option should be called `auto_deduplication` with values `debounce` and `throttle`. `debounce` should the only process the first incoming message and `throttle` the last within a given interval (I know I said timeouts are hard to handle). or maybe even set an `int` value to `auto_deduplication` meaning `0 == debounce` and `> 0 == throttle`? This may be a little to abstract but would make it possible to only use one field.

 > "debounce": an alternative word would be "deduplication" and "auto-dedup" or similar. deduplication is what data scientists know ... debouncing is sth eletronics guys do;)

I'm pro `auto_deduplication`

> "transaction_hash": it needs to be unique .. "hash" suggests that, but also suggest a specific way of computing the value (in the client lib or app) .. which is unnecessary. and "ID" is good enough

I'd rather stick with hash, because you need to compute a hash for this to work. One might think this is part of the [ID generation specs of wamp](https://wamp-proto.org/_static/gen/wamp_latest.html#ids). The specs should just say that choosing the hashing method is done by the implementer. The hashing method MUST be the same for all connected callers and procedure.

> rgd return policy (that is, upon detecting a dup): "return error", "return first", ..?

In my scenatio, skipping the deduplicated messages seem to be the only correct way of handling it. But deferring them would also work. The problem about deferring them is the higher chance of getting a out of memory in the router. Maybe the callee or the router should decide it. For the callee this would mean to have an extra option `auto_deduplication_policy` with values of `cancel` and `defer`.

> feature announcement

auto_deduplication

> in a multi-node setup (using r2r links), the Q "who handles the debounce" becomes: which node? is it the node where the originating caller is connected, or is it the one the actual callee is connected?

does crossbar support multiple routers?

This is in my head how it should work:

```
 ┌───────────┐                         ┌─────────────┐                      ┌──────────────┐
 │           │                         │             │                      │              │
 │  Caller   │                         │   Router    │                      │   Callee     │
 │           │                         │             │                      │              │
 └─────┬─────┘                         └──────┬──────┘                      └───────┬──────┘
       │                                      │                                     │
       │                                      │◄────────────────────────────────────┤
       │                                      │                                     │
       │                                      │ Registration with                   │
       │                                      │ auto_deduplication = 0              │
       │                                      │                                     │
       │                                      │                                     │
       ├─────────────────────────────────────►│                                     │
       │                                      │                                     │
       │ Call with hash 112233                │                                     │
       │                                      │                                     │
       │                            ┌─────────┤                                     │
       │                            │         │                                     │
       │               Add hash to  │         │                                     │
       │               dedup table  │         │                                     │
       │                            │         │                                     │
       │               add(112233)  │         │                                     │
       │                            └────────►├────────────────────────────────────►│
       │                                      │                                     │
       │                                      │ Invoke procedure                    │
       ├─────────────────────────────────────►│                                     │
       │                                      │                                     │
       │ Call with hash 112233                │                                     │
       │                                      │                                     │
       │                            ┌─────────┤                                     │
       │                            │         │                                     │
       │               Found hash in│         │                                     │
       │               dedup table  │         │                                     │
       │                            │         │                                     │
       │               abort()      │         │                                     │
       │                            └────────►│                                     │
       │                                      │                                     │
       │◄─────────────────────────────────────┤                                     │
       │                                      │                                     │
       │ Error 'wamp.deduplicated.messae'     │◄────────────────────────────────────┤
       │                                      │                                     │
       │                            ┌─────────┤ Result                              │
       │               Remove hash  │         │                                     │
       │               from table   │         │                                     │
       │                            │         │                                     │
       │             remove(112233) │         │                                     │
       │                            └────────►│                                     │
       │                                      │                                     │
       │                                      │                                     │
       │                                      │                                     │
```

**Comment by konsultaner** [comment]
While starting to implement a demo. I found out that `throttle` should return an error where as `debounce` should just replace the call within a given timelimit and restart the timelimit. So `debounce` and `defer` is kind of the same IMO.

**Comment by oberstet** [comment]
fantastic feedback! I think this is a lot of "raw material" already .. would be great to get more feedback from other implementors as well ..

and then of course (after a bit more of discussion and rough design), of course implement it in a router, test it, adjust design if required, add it to the spec

> does crossbar support multiple routers?

yeah, since recently, we've open-sourced router-to-router links, which allows you to create multi-node networks, incl. mesh and edge/cloud.

this is currently still "single operator" (all nodes are operated by 1 op, via 1 master node for orchestration)

we _will_ / _want_ to support running router-to-router links between nodes of different operators as well!

**Comment by konsultaner** [comment]
>yeah, since recently, we've open-sourced router-to-router links, which allows you to create multi-node networks, incl. mesh and edge/cloud.
>
>this is currently still "single operator" (all nodes are operated by 1 op, via 1 master node for orchestration)
>
>we will / want to support running router-to-router links between nodes of different operators as well!

are there specs to it? Been wanting to do this forever... I had trouble how to do that. I started with some thoughts but it seems hard to accomplish.

**Comment by oberstet** [comment]
no proper spec yet:( just a bunch of diagrams, here are 2:

* https://github.com/crossbario/crossbar/blob/master/docs-cfx/_static/crossbar-rlinks.pdf
* https://github.com/crossbario/crossbar/blob/master/docs-cfx/_static/cfx-clustering-elements.pdf

**Comment by konsultaner** [comment]
> fantastic feedback! I think this is a lot of "raw material" already .. would be great to get more feedback from other implementors as well ..
>
> and then of course (after a bit more of discussion and rough design), implement it in a router, test it, adjust design if required, add it to the spec

I'm in a desperate need for that feature. Do you mind if I add a pull request with an alpha state, after I implemented a demo in my router?

**Comment by oberstet** [comment]
@konsultaner so I'm +1 on having sth to actually experiment with, even at this very rough stage .. but that would be sth you could also raise as a question on the call later ..

**Comment by konsultaner** [comment]
@oberstet I would love to be more involved in the DAO, but atm there is absolutely no time for that. I'm overloaded with work and being a dad 🙈... sorry

**Comment by konsultaner** [comment]
I implemented it into my router (which will be open souce, hopefully by the end of next year) and it integrates quite well with all other RPC related features. I even found a bug in my router, because I had to generify some code parts. Seems like a good way of making it work. One thing to clear up with this feature:

`auto_deduplication <= 0` == throttle, which means all calls followed by the first call with the same hash will return an error as long as the initial call has not been processed by the callee.

`auto_deduplication > 0` == debounce, which means all calls followed by the first call with the same hash will replace the previous received call until the given due time is reached. The replacement is done silently without an error. When the due time is reached, the router will start to pass the invocation to the callee.

Why the difference in errorhandling?

*Throttle*
- Throttle is used to block the channel
- Scenario: Button click calls a procedure -> don't process every click, tell the user there is something wrong

*Debouce*
- Debouce is used to start a process after no more action is done within a given time period
- Scenario: Server side autocomplete -> only send the search string after the user has approximately finished his input, no need for a user feedback

I'm open for discussion :)

**Comment by Jopie64** [comment]
> Why would I not let the router decide when to debounce?
> 
> 1. opaque payload may make it impossible
> 2. the router would need to store extra state
> 
> Why would I not let the callee decide when to debounce?
> 
> 1. if there are multiple callees registred on the same procedure they would have to sync states
> 2. In my case there are multiple callees but one database -> without syncing the states multiple entries may appear

Out of curiosity:
Why would you not let the calleR decide when to (and handle the) debounce?

Trying to answer it myself:
* Do you expect multiple clients doing the exact same call with the exact same arguments at almost the exact same time? I can't think of such a scenario...
* You don't have control over the source code of the clients? But then, you also wouldn't be able to make it send a hash of the payload.

**Comment by ecorm** [comment]
> Why would you not let the calleR decide when to (and handle the) debounce?

@Jopie64 , consider the use case where the RPC is used to activate a big machine (with the argument ON/OFF). The process of turning the machine on goes through several stages, and it's not allowed for the machine to jump back to the beginning in the activation sequence. It would however be allowed for the machine to be turned back OFF in the middle of the activation sequence (e.g. emergency shutdown).

The caller does not have enough information to filter out the duplicate activate(ON) commands, so it's up to either the router or the callee.

I'm my opinion, the filtering behaviour the OP wants belongs to the business logic of the callee, and not the router, which should remain stateless as much as possible.

If we consider the same use case with multiple callees registered to the same RPC, then things get more complicated and the router needs to get involved in the filtering logic if the callees are not mutually aware of the RPC progression state.

When there are commands in an application that take a long time to complete, what I do is return immediately from the RPC after the state machine has been kicked off. The client can then wait for the completion state to be signalled via a pub/sub event.

If there is only a single callee involved, the filtering feature could be implemented purely in the client library on the callee side, without having to extend the WAMP protocol.

**Comment by konsultaner** [comment]
@Jopie64 I do have that issue sometimes. It seems like some network routers defer data when they reconnect or have WAN issues. My clients clients then seem to try different devices. This ends up in nearly parallel calls. I tried to debounce it on the client side, which helped a lot, but the issue somehow remains. This is one of 3 scenarios that we experience to happen in the real world.

**Comment by konsultaner** [comment]
Rgd `debounce`, maybe there should be an error for each replacement. This would reduce caller complexity to prevent overflowing memory. Streams frameworks that also implemt debounce are not transactional, but WAMP is. So I guess just ignoring a call may cause more errors.

**Comment by oberstet** [comment]
> Why would you not let the calleR decide when to (and handle the) debounce?

with a single caller, a single node (no r2r linked network of nodes), and a single callee, technically, the filtering can be done in any of these 3 locations.

however, I think you hint at sth important with "You don't have control over the source code of the clients". if clients means caller & callees, that only leaves router for filtering.

> the filtering behaviour the OP wants belongs to the business logic of the callee, and not the router

my view: it depends on the _kind_ of "filtering". 

if filtering in an app means eg "only let call pass if `args.amount + kwargs["value"] > 23`, then this is more *business logic* like

if filtering in an app means eg "only let call pass for args & kwargs not seen previously", then this is more *technical deduplication* like

what you hint at though, yes, one design guide we have used in WAMP is: a router should not contain business logic. because it should not be required to understand the app payload (and hence could not apply any business logic).

that means: the filtering the OP wants can only be done in the router based on one of

1. filter based on **hash256(CALL.args | CALL.kwargs)**
2. filter based on **CALL.options.transaction_id|uint256**, where the option value is already set by the caller

> If we consider the same use case with multiple callees registered to the same RPC, then things get more complicated and the router needs to get involved in the filtering logic if the callees are not mutually aware of the RPC progression state.

yes, agreed.

further, with multiple nodes connected in a network (via router-to-router links), the only node capable of 

3. filtering is the *node the caller is directly connected to*.

> maybe there should be an error for each replacement. This would reduce caller complexity to prevent overflowing memory. ...

sorry, I don't understand what you mean .. what replacement? rgd memory, couple of thoughts - if we decide to filter at the (first) router node - :

* the router needs to store a `uint256->obj` map (with optional previous `CALL_RESULT.args,kwargs` or `None` otherwise)
* the "amount" of lookback for the filtering could be defined in terms of a) number of CALL requests b) number of unique CALLs c) seconds walltime
* using b) would be straight forward: keep 2 maps and swap/destroy/create once both are full (contain N items). This can be used to configure the router to limit memory to N then

-----

so in my eyes currently, one feasible and consistent way could be 1. + 2. + 3. above ...

**Comment by konsultaner** [comment]
I think this feature should be seen as an helper to what you would ususally do in the caller anyways. In the frontend world you would us an rxjs or rx-whatever to debounce or throttle calls. This feature just helps to fix networking and synchrinization issues, that cannot be done by the caller. So letting the caller handle when to dedup is what is happening at the moment. To make all other advanced wamp features work in conjunction with auto deduplication, it comes natural that the caller hints the router when to deduplicate. 

I have finished my implementation. The only issue that may appear is a memory overflow. But this very easy to prevent by setting a maximum level of hashes per registration. In the real world there will probably only be one or two hashes in that list anyways. I don't see a big issue here.

This is a flow diagramm for a debounce:

```
┌───────────┐                         ┌─────────────┐                      ┌──────────────┐
│           │                         │             │                      │              │
│  Caller   │                         │   Router    │                      │   Callee     │
│           │                         │             │                      │              │
└─────┬─────┘                         └──────┬──────┘                      └───────┬──────┘
      │                                      │                                     │
      │                                      ◄─────────────────────────────────────┤
      │                                      │                                     │
      │                                      │ Registration with                   │
      │                                      │ auto_deduplication = 300            │
      │                                      │                                     │
      │                                      │                                     │
      ├──────────────────────────────────────►                                     │
      │                                      │                                     │
      │ Call with hash 112233, requestId=1   │                                     │
      │                                      │                                     │
      │                            ┌─────────┤                                     │
      │         schedule that call │         │                                     │
      │         to be executen in  │         │                                     │
      │         300ms              │         │                                     │
      │                            │         │                                     │
      │                            │         │                                     │
      │                            └─────────►                                     │
      │                                      │                                     │
      │                                      │                                     │
      ├──────────────────────────────────────►                                     │
      │                                      │                                     │
      │ Call with hash 112233 after 200ms    │                                     │
      │ with requestId = 2                   │                                     │
      │                            ┌─────────┤                                     │
      │        schedule that call  │         │                                     │
      │        to be executen in   │         │                                     │
      │        300ms               │         │                                     │
      │                            │         │                                     │
      │        replace the former  │         │                                     │
      │        call                └─────────►                                     │
      │                                      │                                     │
      │◄─────────────────────────────────────┤                                     │
      │                                      │                                     │
      │ Error 'wamp.deduplicated.messae' for │                                     │
      │ requestId = 1                        │                                     │
      │                            ┌─────────┤                                     │
      │             another 300ms  │         │                                     │
      │             have passed    │         │                                     │
      │             turn call into │         │                                     │
      │             invocation and │         │     call                            │
      │             send to callee └────────►├────────────────────────────────────►│
      │                                      │     yield                           │
      │◄─────────────────────────────────────┤◄────────────────────────────────────┤
      │ Result for requestId = 2             │                                     │
      │                                      │                                     │
      │                                      │                                     │
      │                                      │                                     │
      │                                      │                                     │
      │                                      │                                     │
      │                                      │                                     │
                                                                                   │
```

@oberstet 
> maybe there should be an error for each replacement. This would reduce caller complexity to prevent overflowing memory

With this I mean, that if you just replace a debounced call in your call stack without sending an error. The caller may have a list of responses it is waiting for. Sending no error will make that list grow infinitly. Thats why responding with an error for every debounce would be a good thing to do.

**Comment by oberstet** [comment]
@konsultaner ok, the diagram is perfect to understand what you've actually implemented, however, I'm afraid from my point of view, what you've implemented doesn't comform to the proposal we've discussed - it's yet a different mechanism

there is no "delay" in the proposal of this issue, and the first call will pass, and any later duplicate ones will be filtered

so a diagram would look different from above

maybe we should really aim for a spec proposal incl. a message sequence diagram first ..

apart from that, I wouldn't implement the feature as in your diagram in crossbar .. it will swamp the router with massive numbers of timers or will require the routing logic to be integrated with an "artificial time/future" abstraction in the event loop / reactor (eg it could be done in Twisted using reactor time, but I wouldn't want to;)

**Comment by konsultaner** [comment]
@oberstet Ok, I understand. The first diagram is a throttel and the second is a debounce. I tried to combine both. Throttle is what I really need. I bet if someone sees the throttle feature, the request of a debounce feature will be very close.

> it will swamp the router with massive numbers of timers or will require the routing logic to be integrated with an "artificial time/future" abstraction in the event loop / reactor (eg it could be done in Twisted using reactor time, but I wouldn't want to;)

This is exactly what happened to me. In Java it is recommended to use a global timer to prevent performance issues. Thats why I use a single threaded timer globally. Thats works quite good for all timing related features. I also had to add a future to handle the debounce. but this was only a little impact. 

I do understand why you don't see debounce in crossbar, but we could make throttle and debounce independent features. This way one could decide wether to implement both or just the timerless throttle version. Throttle and Debounce work with the same options. Its just a matter of the router interpreting the `auto_deduplication` value of the registration options.

**Comment by oberstet** [comment]
as said, I don't assign meaning to the word "debounce" - what I have been talking about in this issue is deduplicate

I should probably emphasize that what you've _implemented_, which I guess is precisely represented in above message sequence chart is _another_ different feature that we can of course also discuss .. it's just quite a different thing compared to what we did discuss. Eg the word "delay" first appeared in your diagram .. and it changes the nature of the feature fundamentally.

It changes the routing logic actually. A call MUST NOT be forwarded to a Callee earlier than T1 ms by the Router, and so on ..

this is a muuuuch different feature, and would deserve a much more thorough analysis IMO

the reason is: at it's core, it requires integration of routing logic and introduces a timing dimension to actual routing itself.

this is a big one, and I am not convinced of the advantages and need when seen in the light of the resulting implications.

anyways, that's just my view. but I'm not interested in opening the timed-routing can at this point ... for crossbar and autobahn

**Comment by konsultaner** [comment]
> as said, I don't assign meaning to the word "debounce" - what I have been talking about in this issue is deduplicate

ok let me clearify this. debounce and throttle and other stream filters/operators are described here: http://reactivex.io/

Well known and widely used frontend frameworks like angular and flutter use streams to organize the dataflow in the frontend and to handle server interaction. WAMP defines a stream as well and thus could also provide stream operators. Some of the most common operators are: timeout, debounce and throttle I guess.

This is a list of operators defined by reactivex: http://reactivex.io/documentation/operators.html#categorized

One stream operator that is supported by WAMP is timeout: https://github.com/wamp-proto/wamp-proto/blob/master/rfc/text/advanced/ap_rpc_call_timeout.md  -> http://reactivex.io/documentation/operators/timeout.html

Debounce means:  only invoke a call from a caller if a particular timespan has passed without it emitting another call
![image](https://user-images.githubusercontent.com/2699538/145966513-549bb5df-6ab4-4a87-b76d-6e348894a49c.png)
Source: http://reactivex.io/documentation/operators/debounce.html

Throttle means: only invoke a call from a caller if no other call has been emitted before it within a particular timespan. For this feature the particular timespan is defined by the computation time of a procedure itself.

Debounce and Throttle [are always named together](http://reactivex.io/documentation/operators/debounce.html). But the major use of both of them is to **deduplicate** messages. Thats why I mixed them up in this feature request. The first flow diagram discribes `throttle`, the second describes `debounce`.

There is a good video describing throttle vs debounce: https://reactive.how/throttletime

> it can keep state and time. so why is there a need for anything more built into the protocol itself?

From the stream perspective the router [merges](http://reactivex.io/documentation/operators/merge.html)  all incoming streams before it passes the invocations to the callees. Because only the router has the merged stream, only the router can process deduplication. The `transaction_id` (or hash) only helps the router to deduplicate, because the router doesn't know anything about the `args` or `kwargs` when using opaque payloads.

in rxjs a WAMP RPC stream would kind of look like that:

```js
const callerStream1 = new CallerStream();
const callerStream2 = new CallerStream();

const routerStream = merge(
  callerStream1,
  callerStream2
);

// output: 
// "call 1 from caller 1 at 0ms", 
// "call 2 from caller 2 at 10ms"
// "call 3 from caller 1 at 11ms",
// "call 4 from caller 1 at 12ms", 
// "call 5 from caller 2 at 16ms"
routerStream.subscribe(val => console.log(val));

// output: 
// "call 1 from caller 1 at 0ms", 
// "call 2 from caller 2 at 10ms"
// "call 5 from caller 2 at 16ms"
// for this feature throttleTime replace the 5ms with the execution time
routerStream.pipe(throttleTime(5)).subscribe(val => console.log(val));

// output: 
// "call 1 from caller 1 at 5ms", 
// "call 2 from caller 2 at 15ms" 
// "call 5 from caller 2 at 20ms"
routerStream.pipe(debounceTime(5)).subscribe(val => console.log(val));
```

I really need the throttle for my current project. I came up with debounce, because I mixed up debounce and throttel. But in the end it was easy to implement and would reduce caller complexity a lot.

**Comment by konsultaner** [comment]
> put into a question: I can simply implement a WAMP procedure for what you call debouncing, which in turn calls the actual target procedure in whatever decouncing fashion you like. IOW: it can keep state and time. so why is there a need for anything more built into the protocol itself?

Because if you use the invocation policy `round_robin` or `random` you would need to sync the timer state between the callees.

**Comment by oberstet** [comment]
rgd the diagram: thanks, yeah, this illustrates the debouncing itself, but it doesn't illustrate what exactly in terms of which calls is debounced, and it doesn't illustrate the actors involved and the messages between those. IOW: it is not a message sequence chart. Is there a complete MSC for this stuff?

> One stream operator that is supported by WAMP is timeout

yes, true, but this is very different. crucially, it applies the timeout to the respective single originating call, and such policy is naturally compatible with `round_robin` etc, and would also be compatible with `partition` or `all` (for parallel calls .. which we not yet have), as it can be done at the originating node. debouncing again only per caller, _not_ across different callers ..

> Because if you use the invocation policy round_robin or random you would need to sync the timer state between the callees.

This is undefined: the above sequence diagram only shows an example with 1 caller and 1 callee. There are multiple ways of _defining_ what should actually happen.

Eg the debouncing being defined at an URI within the whole realm means, the debouncing is covering both multiple originating calls (from different clients) and multiple terminating callees .. and hence it cannot be done at the single caller level. It could also not be done at the originating node level (that is the first node where callers are connected), since there could be multiple originating nodes with callers. That is, for _this_ definition. But as said, there are others as well .. which is it?

**Comment by konsultaner** [comment]
@oberstet I tried my best to visualize it. My 2 examples are with multiple router nodes as far as I understand how it might work and invocation policy `round_robin`. All shown timestamps assume we are on the `01.01.1970 00:00:00.000`.

Throttle:
![Deduplicate-Throttle drawio](https://user-images.githubusercontent.com/2699538/146520816-0211743d-a842-4f5c-beec-71f227eaf7cd.png)

Debounce:
![Deduplicate-Debounce drawio](https://user-images.githubusercontent.com/2699538/146520853-259f1822-b672-4d67-8b73-ff0b20e8b619.png)

This is the [draw.io](https://app.diagrams.net/) file if someone want's to fix it or add something;
[Deduplicate.zip](https://github.com/wamp-proto/wamp-proto/files/7733972/Deduplicate.zip)

**Comment by oberstet** [comment]
@konsultaner first of all: awesome! I really think this way of definition and presentation is really useful because it is almost enough to "run it" (a complete spec). also perfect to include multiple clients and multiple router nodes in the set of actors shown.  and respective messaging scenarios.

it seems some few messages are still missing in the design: eg "store hash and pass to other node". in this design, a respective WAMP message is needed then from router 1 to 2. also, is the callee on router 1 in the example?

if so, why does router 2 need the hash of the first call? if the second call with same hash arrives at router 2, it could still be forwarded to router 1 (which has the callee connected), and router 1 would do the dedup ...

**Comment by oberstet** [comment]
the other thing your last comment made me think of: couldn't we split the design problem in 2 parts? as in:

1. specify "wamp action hash" (eg for CALL) at the message level
2. specify how a node behaves wrt "wamp action hash":
  a) return error or first result for all but the first action (and keep a max depth D of hashes per URI)
  b) only execute the N-th action with same hash within M ms
  c) ...

in my perspective 1. + 2a. would be useful and doesn't introduce "routing target timing" into the router

1. + 2b. is what you want as far as I understand, right? as said, this introduces a much more fundamental change into the whole router semantics: "routing target timing". 

in my eyes, rgd 1.+ 2b. would therefor actually better be discussed by first discussing "routing target timing" concept .. and only then within that context, discuss one example using this, which would be 1. + 2b.

not sure if I was able to express what I mean .. what do you think?

**Comment by oberstet** [comment]
> because it is almost enough to "run it" (a complete spec).

expanding on this, the way I think of Message Sequence Charts (MSC) for WAMP at this detail level is:

if we had a complete set of MSCs for WAMP, and if those MSCs where in machine readable format (eg some JSON), then this _spec_ could be started as a fully functional WAMP router - given a very smart MSC meta interpreter

IOW: the spec is the program and can be run to get a live instance. _if_ the MSCs are complete that is;)

**Comment by konsultaner** [comment]
I will answer to this in detail later. For a quick response: 

> couldn't we split the design problem in 2 parts?

That is brilliant! Great idea. The `transaction_hash` may be used for other features as well. One could basically add a other stream operators apart from `throttle` and `debounce` if they are needed.

I'm +1 for splitting the request into 2 parts. 

> in my perspective 1. + 2a. would be useful and doesn't introduce "routing target timing" into the router

I know @oberstet doesn't want to implement a timer based routing into crossbar, but does crossbar need to be fully featured? Is a wamp feature only accepted if it eventually ends up as a part of crossbar? I implemented it with no performance issues at all. And it runs like charm in even in production. My router implementation also has totally different targets then crossbar.

> expanding on this, the way I think of Message Sequence Charts (MSC) for WAMP at this detail level is:

From now on I will only add feature requests as MSC. I love them, but thought it might be to hard to understand if I only present things with MSCs.

> if we had a complete set of MSCs for WAMP, and if those MSCs where in machine readable format (eg some JSON), then this spec could be started as a fully functional WAMP router - given a very smart MSC meta interpreter

You are talking about some [EBNF](https://en.wikipedia.org/wiki/Extended_Backus%E2%80%93Naur_form) like MSC? Some kind of generic format. I think this is a good idea to test features. Even performance may not be a problem in the end. Payload handling could be generalized too. I mean you would just have to tell your subsystem how to zero-copy payloads and set some state handlers, i.e. max_memory or max_stack_depth. 

@oberstet You should start a new issue about WMSD (WAMP message sequence description) ;)

**Comment by konsultaner** [comment]
@oberstet rgd MSCs. Maybe looking at WAMP as a stream might be a good point to start at. Streams in reativex are usually developed by using a [DSL](https://en.wikipedia.org/wiki/Domain-specific_language) that defines all necessary stream operators. My [dart implementation](https://github.com/konsultaner/connectanum-dart) is based on streams and stream operators too. It is kind of a proof that it should work. At least for clients 😄

**Comment by oberstet** [comment]
sounds cool! fwiw, what I meant with "this spec could be started as a fully functional WAMP router" is: is the spec executable?

IOW, the interesting question is: given the spec in your DSL file `.dsl`, can you do `dslinterpreter run example.dsl` to start a live WAMP router?

Because if you can, why would you develop a WAMP router program manually anyways?

Of course, this is a pretty unrealistic expectation;) But in principle. So it is more academic IMO .. it could make up for a great master thesis or whatever though ...

> You should start a new issue about WMSD(WAMP message sequence description) or MSDL(message sequence description language) to be more generalized ;)

wrote something here https://github.com/wamp-proto/wamp-proto/issues/394

 > I know @oberstet doesn't want to implement a timer based routing into crossbar, but does crossbar need to be fully featured? Is a wamp feature only accepted if it eventually ends up as a part of crossbar? I implemented it with no performa

yes, true! I mean, right now, I can't see how to justify the effort to add timing into crossbar - BUT: that of course doesn't mean we could specify it in WAMP AP, and you can of course implement and feature-announce this in your router!

so great, I am also +1 on splitting this issue then, into the `transaction_hash` part (the WAMP messages ..), and into a 2nd part along router behavior building on that, like `debounce`.

so 2 AP features: `transaction_hash` and `debounce`

**Comment by oberstet** [comment]
Alright, as discussed above, I'm adding experimental support for `transaction_hash` in AutobahnPython:

* **autobahn.wamp.types.CallOptions**
* **autobahn.wamp.types.PublishOptions**

Should the `transaction_hash` be forwarded in

* **autobahn.wamp.types.CallDetails** to the Callee(s)
* **autobahn.wamp.types.EventDetails** to Subscribers

in case whatever logic is applied to `transaction_hash` router-side is allowing the Call or Event to pass?

![grafik](https://user-images.githubusercontent.com/233340/153691373-10fc92c4-1bae-497a-8be9-e1e0320b5a71.png)
![grafik](https://user-images.githubusercontent.com/233340/153691398-01dd151e-256a-4389-a407-c9ac09f6cf65.png)

**Comment by oberstet** [comment]
fwiw, I seem to have a use case for "deduplication" use of transaction_hash myself now;) which means, that's what I would add (experimental) to Crossbar.io after the ABPy merge.

then, while thinking ahead of implementation in the router for dedup: this can be misused by clients for attacks. eg a client might send app payload and a transaction hash which actually do not match, to prevent others from sending a payload that would hash to the transaction hash. "App-level denial-of-service" ...

**Comment by konsultaner** [comment]
@oberstet thats an interesting issue. Maybe the hash should have some predefined crypto behind it. This should not be too hard to solve.

**Comment by oberstet** [comment]
when a client sends application payload with an "app transaction" and a `transaction_hash` that does not match the transaction, this can be used by the (bad) client for:

* rendering deduplication ineffective, as a 2nd duplicate can always be sent using a fake `transaction_hash`, and such an Event/Call will be processed by the router
* denying another client to issue a Event/Call with actual matching app payload / `transaction_hash` by consuming the `transaction_hash` earlier

IOW: the router must trust a single client, and the router behavior for all clients will be affected.

If the `transaction_hash` would be computed by the router from `args|kwargs`, the problem disappears.

However, "computing transaction_hash from args|kwargs" can either

* be done without qualification, over the _complete_ `args|kwargs` OR
* only portions of those

The latter is obviously much preferable. But how?

One way would be: have 2 modes for the feature.

1. The client provides `transaction_hash` (permitted by the client role on the respective URI and action) OR
2. The client provides `transaction_hash_args: list|int` and `transaction_hash_kwargs: list|int`

In the latter case, the positions of items within `args` and the names of items within `kwargs` that are supposed to collect the data over which the router computes the hash are specified.

There are 2 problems with that:

* given `transaction_hash_args` and `transaction_hash_kwargs`, how do be serialize that to bytes to compute the hash in a way that is precisely defined? eg serializer formats usually do not specify a bijection between values and bytes. IOW: this is implementation dependent
* a client can still misuse the feature, as it can direct the router to take items for hash computation the app didn't plan for

So maybe this points to an actual solution. Consider the following potential router config (just for discussion, not for the feature spec):

```json
{
   "workers": [
        {
            "type": "router",
            "realms": [
                {
                    "name": "realm1",
                    "roles": [
                        {
                            "name": "anonymous",
                            "permissions": [
                                {
                                    "uri": "com.example.uri1",
                                    "match": "exact",
                                    "allow": {
                                        "call": true,
                                        "register": true,
                                        "publish": true,
                                        "subscribe": true
                                    },
                                    "disclose": {
                                        "caller": true,
                                        "publisher": true
                                    },
                                    "deduplicate": {
                                        "caller": true,
                                        "publisher": true
                                    },
                                    "transaction_hash": {
                                        "args": [2, 3],
                                        "kwargs": ["price", "customer"]
                                    },
                                    "cache": true
                                }
                            ]
                        }
                    ]
                }
            ]
        }
    ]
}
```

**Comment by oberstet** [comment]
@konsultaner also, now that I started to write some code, more questions pop up;) Would be interested to know:

* How do you configure throttle based usage of `transaction_hash` for URI-actions in your router (in a mode where `transaction_hash` is computed and sent by the client)?
* Is throttling per URI-action (in a realm), or per client (depending on role)?

**Comment by oberstet** [comment]
ok, so for completeness, I've added `transaction_hash` to AutobahnPython for the *client-supplied transaction hash mode* in these WAMP messages:

* `Call.Options.transaction_hash | string`
* `Publish.Options.transaction_hash | string`

and for the *client-supplied transaction hash mode* and the *router-computed transaction hash mode* in these WAMP messages:

* `Invocation.Details.transaction_hash | string`
* `Event.Details.transaction_hash | string`

**Comment by konsultaner** [comment]
> @konsultaner also, now that I started to write some code, more questions pop up;) Would be interested to know:
> 
> * How do you configure throttle based usage of `transaction_hash` for URI-actions in your router (in a mode where `transaction_hash` is computed and sent by the client)?
> * Is throttling per URI-action (in a realm), or per client (depending on role)?

@oberstet I don't quite understand the term `uri-action`. What I did was this:

- The client decides on how the `transaction_hash` is created
- The client creates the hash using either values of the kwargs, args, session_id
    - To prevent a DOS attack I could see adding some crypto from the authentication process to be added, i.e. SCRAM or CRA? Havn't thought about that too much, but seems like a good way. -> I'll be adding an idea in the next days.
- The router hadles dedup only within a realm or less.
- If the client wants to prevent dedup only on its own session, include session id in the hash creation
- If the client wants to prevent dedup through out the realm don't session id
- The router is only passively handling the dedup, since dedup is a part of the business logic in my cases, this seems logical

> when a call is deduped at the outgoing node, that is actually the invocation is prohibited, an Error can be returned to the Caller

In my implementation I return a `wamp.error.auto_deduplication` error in this case. In production this makes the client code less complex.

> in any case, with PubSub, we cannot return an Error even for acknowledged publications! when the client is authorized to publish, it will get back a publication ID with acknowledged publications - and then the event is dropped at each outgoing node ... mmmh

I havn't implemented dedup on PubSub yet.

>```json
>"deduplicate": {
>    "publish": true
>},
>"transaction_hash": {
>    "kwargs": ["rx_sha256"]
>}
>```

I haven't implemented router side dedup, because I see it as a part of the business logic. But in your case it seem to make sence. 

> If the transaction_hash would be computed by the router from args|kwargs, the problem disappears.

IMO this would make it even harder to prevent DOS attacks.? If for example the router computes the `transaction_hash` from a `kwargs` field. The client could also send the `kwargs` with the blocking values over and over again. If the router had the chance to validate the hash against something. This would stop this DOS vector at least for session wide deduplication.

I'm not sure how to prevent DOS attack for a realm wide dedup?

What do I use dedup for:

- auto complete fields
- long running processes only starting once (i.e. an accounting process)
- processing that is dependant on states stored to the database
    - Sometimes hardware routers seem to store outgoing messages and send them right after another -> killed some of my logics

**Comment by oberstet** [comment]
>  I don't quite understand the term uri-action

that's just the set `WAMP URIs  x WAMP Actions (Publish, Subscribe, Call, Register)` over which permissions (allow|deny) are defined

> To prevent a DOS attack I could see adding some crypto from the authentication process to be added

I guess you mean *signing* the messages by the publisher/caller, including the transaction_hash?

My question would be: what should that add for the router regarding trust it puts into the client, as the client is presumaly already authenticated (at the WAMP session level) and authorized (at the WAMP URI-Action level)?

> The router hadles dedup only within a realm or less.

Rgd "only with a realm": agreed! and this is important.

"or less": this is also crucial. I would argue for, given this example (in Crossbar.io config syntax) permission:

```
{
    "uri": "com.example.uri1.",
    "match": "prefix",
    "allow": {
        "call": true,
        "register": true,
        "publish": true,
        "subscribe": true
    },
    "deduplicate": {
        "call": true,
        "publish": true
    }
}
```

that the deduplication with a client supplied `transaction_id` is applied to all Calls and Publications to any URI matching the prefix `com.example.uri1.`, on the given Realm R.

thus, after a Publish to `com.example.uri1.topic1` with `transaction_hash=H1`

1. another Publish or Call with same H1 to any URI `com.example.uri1.*` will fail?
2. any Publish or Call to an URI that does match a different pattern can use the same H1?
3. any Publish or Call to an URI matching the pattern, but without a `transaction_id` will succeed?
4. anything on a different Realm R2 is totally independent (the same H1 could be used)

> The router is only passively handling the dedup

Not sure I understand what you mean. The router must drop (not deliver) any duplicate Call or Publish as to above rules. Otherwise `transaction_id` doesn't have an effect on routing behavior and could be simply transported as app payload.

> I haven't implemented router side dedup, because I see it as a part of the business logic. 

No, it's a routing feature. Deduplication. It is just that the value on which the deduplication happens may be computed client-side or router-side.

With the client computing `transaction_hash`, the router is still in charge of dropping (not delivering) the Call or Publish (as an Invocation or Event).

> If for example the router computes the transaction_hash from a kwargs field. The client could also send the kwargs with the blocking values over and over again.

Yes, this is actually true.

If we let the app specify dedup via router config, and let the router enforce that, any client authenticated and authorized, will be able to use any args|kwargs, and by that denying other clients from reusing the same args|kwargs.

effectively, all clients in a realm operated under one ruleset, and hence the implicitly trust each other.

mmh .. I think this is a more general, and different problem ...

**Comment by konsultaner** [comment]
> > I haven't implemented router side dedup, because I see it as a part of the business logic.
>
> No, it's a routing feature. Deduplication. It is just that the value on which the deduplication happens may be computed client-> side or router-side.

I meant the `transaction_hash` creation is done only on the client side in my implementation. I definitly have implemented this on the router side. Otherwize it wouldn't be possible 😅 I don't want the router so have to look into the payload, because I use opaque payloads for the router kind of in general. I need this feature to make zero-copy payload transfer between all the clients possible.

> My question would be: what should that add for the router regarding trust it puts into the client, as the client is presumaly already authenticated (at the WAMP session level) and authorized (at the WAMP URI-Action level)?

This would only add a little "security" to it and only if the client wants a session wide dedup. For a realm wide dedup this would do nothing. But to be honest, WAMP has a DOS problem in general. DOS prevention in WAMP seems to be a very hard to solve problem. Does crossbar have a rock solid DOS prevention?

But as a matter of fact, this adds an additional DOS vector. The good thing is, all DOS issues only belong to already authenticated (trusted or at least known) clients.

Maybe an invocation policy would help. i.e. round_robin. So if 2 clients keep competing for a call the router could block the last caller for a defined amount of time. Like 10ms or something very low. Just a quick thought, nothing precise.

> thus, after a Publish to com.example.uri1.topic1 with transaction_hash=H1
>
> - [ ] another Publish or Call with same H1 to any URI com.example.uri1.* will fail?
> - [ ] any Publish or Call to an URI that does match a different pattern can use the same H1?
> - [ ] any Publish or Call to an URI matching the pattern, but without a transaction_id will succeed?
> - [ ] anything on a different Realm R2 is totally independent (the same H1 could be used)?

From my perspective and how I need this feature:

- [ ] another Publish or Call with same H1 to any URI com.example.uri1.* will fail?
- [x] any Publish or Call to an URI that does match a different pattern can use the same H1?
- [x] any Publish or Call to an URI matching the pattern, but without a transaction_id will succeed?
- [x] anything on a different Realm R2 is totally independent (the same H1 could be used)?

>```json
>{
>    "uri": "com.example.uri1.",
>    "match": "prefix",
>    "allow": {
>        "call": true,
>        "register": true,
>        "publish": true,
>        "subscribe": true
>    },
>    "deduplicate": {
>        "call": true,
>        "publish": true
>    }
>}
>```

That's a good point, I should add this to my implementation too.
